cmd/stream-commp/main.go

```diff
@@ -1,64 +1,198 @@
 package main
 
 import (
+	"bufio"
+	"encoding/binary"
+	"fmt"
 	"io"
+	"io/ioutil"
 	"log"
 	"os"
 
 	commcid "github.com/filecoin-project/go-fil-commcid"
 	commp "github.com/filecoin-project/go-fil-commp-hashhash"
+	"github.com/ipfs/go-cid"
+	cbor "github.com/ipfs/go-ipld-cbor"
+	"github.com/mattn/go-isatty"
 	"github.com/pborman/options"
 )
 
+type CarHeader struct {
+	Roots   []cid.Cid
+	Version uint64
+}
+
+func init() {
+	cbor.RegisterCborType(CarHeader{})
+}
+
+const BufSize = ((4 << 20) / 128 * 127)
+
 func main() {
 
 	opts := &struct {
-		TargetPieceSize uint64       `getopt:"-t --target-piece-size  Optional target size, larger than the original input one would like to pad to"`
-		Help            options.Help `getopt:"-h --help               Display help"`
+		DisableStreamScan bool         `getopt:"-d --disable-stream-scan If set do not try to scan the contents of the stream for a potential .car stream (much faster)"`
+		PadPieceSize      uint64       `getopt:"-p --pad-piece-size      Optional target power-of-two piece size, larger than the original input, one would like to pad to"`
+		Help              options.Help `getopt:"-h --help                Display help"`
 	}{}
 
 	options.RegisterAndParse(opts)
 
-	log.Println("Reading from STDIN...")
+	if isatty.IsTerminal(os.Stdin.Fd()) || isatty.IsCygwinTerminal(os.Stdin.Fd()) {
+		log.Println("Reading from STDIN...")
+	}
+
 	cp := new(commp.Calc)
-	n, err := io.Copy(cp, os.Stdin)
-	if err != nil {
-		log.Fatal(err)
+	streamBuf := bufio.NewReaderSize(
+		io.TeeReader(os.Stdin, cp),
+		BufSize,
+	)
+
+	var streamLen, blockCount int64
+	var brokenCar bool
+	var carHdr *CarHeader
+
+	if !opts.DisableStreamScan {
+		// pretend the stream is a car and try to parse it
+		// everything is opportunistic - keep descending on every err == nil
+		if maybeHeaderLen, err := streamBuf.Peek(10); err == nil {
+			if hdrLen, viLen := binary.Uvarint(maybeHeaderLen); viLen > 0 && hdrLen > 0 {
+				actualViLen, err := io.CopyN(ioutil.Discard, streamBuf, int64(viLen))
```

Comment:
Since `bufio.Reader` is a [BytesReader](https://golang.org/pkg/bufio/#Reader.ReadByte) maybe [ReadUvarint](https://golang.org/pkg/encoding/binary/#ReadUvarint) can be leveraged instead of Peek+Uvarint+copyDiscard combo.
nbd.

Reply:
I did that originally, but then realized `ReadUvarint` will hide `actualViLen` from me so if the frame RLE is broken just right I won't be able to keep track of the correct `streamLen`

@jsign am I missing an obvious way to do this right/easier?

Reply:
Oh, I see. A bit unfortunate that doesn't return the read size. 

Dunno if would be much easier, but maybe having a `TeeReader` with a goroutine that's just counting bytes that are being read as to keep track of `streamLen`. But since `streamLen` is used in `Fatalf` errors here (which I think it's useful), that might be a bit cumbersome anyway.

I think the `Peek` way it's pretty clear anyway and simplifies if you need `streamLen` at the error levels. If it would be only used in the final output, then maybe the `TeeReader` idea would be better to simplify nestings.